### PR TITLE
AudioSystem Project

### PR DESCRIPTION
**Pull request by rf152**
- Added AudioSystem project
- Added CREDITS.md
- Added badges to README

### CODE REVIEW

File: ApiGateway/Controllers/SignalController.cs [hunk 1]
```diff
@@ -51,15 +46,7 @@ public async Task<IActionResult> SaveSignal(string name, SignalInput input)
...
-            Output = input.OutputState switch
-            {
-                "danger" => "danger",
-                "caution" => "caution",
-                "clear" => "clear",
-                "shunt" => "shunt",
-                "system" => "system",
-                _ => throw new BadHttpRequestException("Invalid Output State"),
-            },
+            Output = input.OutputState
```
Comment (girlpunk) [inline ApiGateway/Controllers/SignalController.cs:49]: To match style guidelines, this should have a comma at the end of the line

File: ApiGateway/Controllers/SignalController.cs [hunk 2]
```diff
@@ -41,7 +36,7 @@ public ActionResult<Signal> GetSignal(string name)
     }
 
     [HttpPut("{name}", Name = "SaveSignal")]
-    public async Task<IActionResult> SaveSignal(string name, SignalInput input)
+    public async Task<IActionResult> SaveSignal(string name, SignalPut input)
```
Comment (girlpunk) [inline ApiGateway/Controllers/SignalController.cs:39]: Should technically be `SaveSugnalAsync`

File: ApiGateway/Models/SignalOutput.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+namespace MRS.ApiGateway.Models;
+
+using System.Runtime.Serialization;
+
+public enum SignalOutput
```
Comment (girlpunk) [inline ApiGateway/Models/SignalOutput.cs:6]: Add `[Serializable]` to get rid of the warning, can go onto anything that's going to/from JSON or MQTT

File: ApiGateway/Models/SignalOutput.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+{
+    //danger|caution|clear|shunt|noaspect
+    [EnumMember(Value = "danger")] Danger = 0,
+    [EnumMember(Value = "caution")] Caution = 1,
+    [EnumMember(Value = "clear")] Clear = 2,
+    [EnumMember(Value = "shunt")] Shunt = 3,
+    [EnumMember(Value = "noaspect")] Noaspect = 4,
```
Comment (girlpunk) [inline ApiGateway/Models/SignalOutput.cs]: `NoAspect`

File: ApiGateway/Models/SignalOutput.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    [EnumMember(Value = "unknown")] Unknown = 5
```
Comment (girlpunk) [inline ApiGateway/Models/SignalOutput.cs:14]: Missing trailing comma

File: ApiGateway/Mqtt/PointStateMessageHandler.cs [hunk 6]
```diff
@@ -1,8 +1,7 @@
 namespace MRS.ApiGateway.Mqtt;
```
Comment (girlpunk) [inline ApiGateway/Mqtt/PointStateMessageHandler.cs:1]: Should MQTT be capitalised here?

File: ApiGateway/Mqtt/PointStateMessageHandler.cs [hunk 7]
```diff
@@ -72,35 +76,23 @@ private void HandleSystem(string name)
...
         var message = JsonSerializer.Deserialize<InputMessage>(GetMessagePayload());
         if (message == null)
-        {
-            _logger.LogDebug("Actually, it's null");
             return;
-        }
 
         PointInput inputState;
 
         if (message.normal > 0 && message.reverse == 0)
```
Comment (girlpunk) [inline ApiGateway/Mqtt/PointStateMessageHandler.cs:85]: Not sure what the scan is complaining about, wouldn't worry too much

File: ApiGateway/Program.cs [hunk 8]
```diff
@@ -20,9 +26,20 @@
 builder.Services.AddSingleton<Cache<Signal>>();
 builder.Services.AddSingleton<Cache<Tracksection>>();
 
-// TODO: Connection details should be loaded from a config provider
 builder.Services
-    .AddSingleton(static services => new MQTTService(services.GetRequiredService<ILogger<MQTTService>>(), "webapi", "192.168.255.6", 1883, "richard", "password", "signalling/", true, true))
+    .AddSingleton(
```
Comment (girlpunk) [inline ApiGateway/Program.cs:30]: Put the defaults into `appsettings.json`, and use environment variables or an `appsettings.Development.Json` for overrides. the Lines 34-38 then become `builder.Configuration.GetValue<string>("MQTT_HOST")` and similar

Also may want to consider an `MQTT` section inside the config (`var MQTTConfiguration = builder.Configuration.GetSection("MQTT");`)

File: AudioSystem/AudioSystem.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (girlpunk) [inline AudioSystem/AudioSystem.py:1]: Any particular reason this is done in Python?

File: AudioSystem/AudioSystem.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+def main():
+    """! Main entry point"""
+
+    logging.info("Booting")
+    mqtt.init()
+    audio.init()
+    while True:
+        audio.loop()
+
+main()
```
Comment (girlpunk) [inline AudioSystem/AudioSystem.py:30]: Allow importing as a module by doing
```python
if __name__ == "__main__":
    main()
```

File: Mqtt/BaseMessageHandler.cs [hunk 11]
```diff
@@ -7,11 +7,11 @@ namespace MRS.Mqtt;
 
 public abstract class BaseMessageHandler
 {
-    protected class MessageWrapper
+    protected struct MessageWrapper
     {
-        [JsonPropertyName("username")] public string Username;
-        [JsonPropertyName("source")] public string Source;
-        [JsonPropertyName("payload")] public Object Payload;
+        [JsonPropertyName("username")] public string username { get; set; }
```
Comment (girlpunk) [inline Mqtt/BaseMessageHandler.cs]: As the scan says, these should have PascalCase names

File: Mqtt/BaseMessageHandler.cs [hunk 12]
```diff
@@ -7,11 +7,11 @@ namespace MRS.Mqtt;
...
+        [JsonPropertyName("source")] public string source { get; set; }
+        [JsonPropertyName("payload")] public object payload { get; set; }
     }
 
     protected MqttApplicationMessageReceivedEventArgs _messageEvent;
```
Comment (girlpunk) [inline Mqtt/BaseMessageHandler.cs:18]: Either make these private, or properties

Comment (girlpunk) [inline Mqtt/BaseMessageHandler.cs:18]: Scan is correct, you need to either declare that the fields can be nullable (and handle null values where necessary), or ensure they have a value when created

File: Mqtt/Messages/Enums/PointOverrideInput.cs [hunk 13]
```diff
@@ -5,7 +5,8 @@ namespace MRS.Mqtt.Messages.Enums;
 public enum PointOverrideInput
 {
     [EnumMember(Value = "system")] System = -1,
```
Comment (girlpunk) [inline Mqtt/Messages/Enums/PointOverrideInput.cs:7]: As above, add `[Serializable]`

File: ApiGateway/Models/Signal.cs [hunk 14]
```diff
@@ -1,10 +1,12 @@
 namespace MRS.ApiGateway.Models;
 
+using MRS.Mqtt.Messages.Enums;
+
 [Serializable]
 public class Signal
 {
     // TODO: These should probably be enums
```
Comment (girlpunk) [inline ApiGateway/Models/Signal.cs]: Can remove the todo :)

File: ApiGateway/Program.cs (hunk 8 above)
Comment (girlpunk) [inline ApiGateway/Program.cs:30]: Technically this should use `services.GetRequiredService<IConfiguration>()`, as the builder might not be available at the time

**Review by girlpunk — CHANGES_REQUESTED** [review]

File: ApiGateway/Models/SignalOutput.cs (hunk 5 above)
Comment (rf152) [reply ApiGateway/Models/SignalOutput.cs:14]: I had the trailing comma on, and then the style guidlines complained...

File: ApiGateway/Mqtt/PointStateMessageHandler.cs (hunk 6 above)
Comment (rf152) [reply ApiGateway/Mqtt/PointStateMessageHandler.cs:1]: No, deliberately done as it is as then you can spot the change in word if it goes on for more than one (As in ApiGateway)

File: AudioSystem/AudioSystem.py (hunk 9 above)
Comment (rf152) [reply AudioSystem/AudioSystem.py:1]: Cross compatibility. Can't find a c# lib that will allow the manipulation of the audio easily and is cross platform.

File: ApiGateway/Program.cs (hunk 8 above)
Comment (rf152) [reply ApiGateway/Program.cs:30]: Will do in the future, just not in this instance - this merge is mostly to get the audiosystem project in.

File: ApiGateway/Controllers/SignalController.cs (hunk 1 above)
Comment (rf152) [reply ApiGateway/Controllers/SignalController.cs:49]: As below, added a comma and the code inspector complained.

File: ApiGateway/Controllers/SignalController.cs (hunk 2 above)
Comment (rf152) [reply ApiGateway/Controllers/SignalController.cs:39]: Agreed, will do in future (see note about this mostly being about putting the AudioSystem project in)

File: ApiGateway/Models/SignalOutput.cs (hunk 3 above)
Comment (rf152) [reply ApiGateway/Models/SignalOutput.cs:6]: Done

**Pull request merged**